pyiceberg/io/pyarrow.py

```diff
@@ -638,10 +634,160 @@ def visit_or(self, left_result: pc.Expression, right_result: pc.Expression) -> p
+        self._handle_null_unmentioned(term)
+        self._handle_nan_unmentioned(term)
+
+    def visit_less_than_or_equal(self, term: BoundTerm[Any], literal: Literal[Any]) -> None:
+        self._handle_null_unmentioned(term)
+        self._handle_nan_unmentioned(term)
+
+    def visit_starts_with(self, term: BoundTerm[Any], literal: Literal[Any]) -> None:
+        self._handle_null_unmentioned(term)
+        self._handle_nan_unmentioned(term)
+
+    def visit_not_starts_with(self, term: BoundTerm[Any], literal: Literal[Any]) -> None:
+        self._handle_null_unmentioned(term)
+        self._handle_nan_unmentioned(term)
+
+    def visit_true(self) -> None:
+        return
+
+    def visit_false(self) -> None:
+        return
+
+    def visit_not(self, child_result: pc.Expression) -> None:
+        return
+
+    def visit_and(self, left_result: pc.Expression, right_result: pc.Expression) -> None:
+        return
+
+    def visit_or(self, left_result: pc.Expression, right_result: pc.Expression) -> None:
+        return
+
+
+def _get_null_nan_refs(
+    expr: BooleanExpression,
+) -> tuple[Set[BoundReference[Any]], Set[BoundReference[Any]], Set[BoundReference[Any]], Set[BoundReference[Any]]]:
+    """Collect the bound terms categorized by having at least one is_null or is_not_null in the expr and the remaining."""
+    collector = _CollectNullNaNUnmentionedTermsFromExpression()
+    boolean_expression_visit(expr, collector)
+
+    def _downcast_term_to_reference(bound_terms: Set[BoundTerm[Any]]) -> Set[BoundReference[Any]]:
+        """Handle mypy check for BoundTerm -> BoundReference."""
+        bound_refs: Set[BoundReference[Any]] = set()
+        for t in bound_terms:
+            if not isinstance(t, BoundReference):
+                raise ValueError("Collected Bound Term that is not reference.")
+            else:
+                bound_refs.add(t)
+        return bound_refs
+
+    null_unmentioned_bound_refs: Set[BoundReference[Any]] = _downcast_term_to_reference(collector.null_unmentioned_bound_terms)
+    is_null_or_not_bound_refs: Set[BoundReference[Any]] = _downcast_term_to_reference(collector.is_null_or_not_bound_terms)
+    nan_unmentioned_bound_refs: Set[BoundReference[Any]] = _downcast_term_to_reference(collector.nan_unmentioned_bound_terms)
+    is_nan_or_not_bound_refs: Set[BoundReference[Any]] = _downcast_term_to_reference(collector.is_nan_or_not_bound_terms)
+
+    return null_unmentioned_bound_refs, nan_unmentioned_bound_refs, is_null_or_not_bound_refs, is_nan_or_not_bound_refs
+
+
 def expression_to_pyarrow(expr: BooleanExpression) -> pc.Expression:
     return boolean_expression_visit(expr, _ConvertToArrowExpression())
 
 
+def _expression_to_complimentary_pyarrow(expr: BooleanExpression) -> pc.Expression:
+    """Complimentary filter conversion function of expression_to_pyarrow.
+
+    Could not use expression_to_pyarrow(Not(expr)) to achieve this complimentary effect because ~ in pc.Expression does not handle null.
```

Comment:
nit:

```suggestion
def _expression_to_complementary_pyarrow(expr: BooleanExpression) -> pc.Expression:
    """Complementary filter conversion function of expression_to_pyarrow.

    Could not use expression_to_pyarrow(Not(expr)) to achieve this complementary effect because ~ in pc.Expression does not handle null.
```